### ISSUE TITLE
Can CPU resource scheduling be scheduled through Cgroup?

### ISSUE THREAD

**Issue by wayyoungboy** [ray-project/ray#10037]
The current resource scheduling restrictions rely on programmers to manually set quantitative indicators.

**Comment by rkooo567** [comment]
You mean restrict the machine resource usages using cgroup? This is kind of done for actors, but not for tasks. If I misunderstood, can you elaborate the implementation idea a little bit more?

**Comment by wayyoungboy** [comment]
For example, the current allocation of resources depends on the programmer's manual allocation. I mark a task as a unit of CPU, but in the actual call process, the CPU used by the program exceeds 100%. Perhaps ray can limit the resource usage of this task through Cgroup.

**Comment by wuisawesome** [comment]
So the short answer is no: since Ray is independent of any containerization technology (including Linux).

Whether or not we should attempt to enforce allocations via cgroups when possible is potentially worth discussing though. 

cc @ yiranwang52

**Comment by yiranwang52** [comment]
I think the options are：
1. Ray itself manages cgroup per actor/task, like YARN. This will take a lot of effort to implement. Hard to handle corner cases like restart, cleanup, or when users launches ray itself into a container.
2. Ray doesn't manage cgoups/containers at all. Let users leverage existing container orchestration frameworks like k8s, and address this need by right-sizing the containers and using constraints.
3. Tight integration with docker or similar container technology.
4. Tight integration with k8s.

There are discussions about 3 and 4, and will be worked on later. CC @simon-mo

**Comment by rkooo567** [comment]
I believe 3 and 4 are the right way to go. Btw, afaik, we are already using cgroup for actor tasks to constraint memory usage

**Comment by zhe-thoughts** [comment]
@ericl @rkooo567 Could you confirm whether we are already using cgroup for workers? Thanks

**Comment by rkooo567** [comment]
I don't think so. I think we only use cgroup for actors to contraint memory usage IIRC. I believe this is the RFC that can solve this problem; https://github.com/ray-project/ray/issues/14077, but I don't know the current progress

**Comment by ericl** [comment]
Nope. That would basically amount to the containerization support, which
would provide that for free.

**Comment by cszhu** [comment]
Hello! This P3 issue has seen no activity in at least a year. It will be closed in 2 weeks as part of ongoing cleanup efforts.

Please remove the `pending-cleanup` label if you believe this issue should remain open.

Thanks for contributing to Ray!
